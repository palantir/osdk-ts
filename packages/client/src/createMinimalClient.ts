/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createSharedClientContext } from "@osdk/shared.client.impl";
import type { Logger } from "pino";
import type {
  ClientCacheKey,
  MinimalClient,
  MinimalClientParams,
} from "./MinimalClientContext.js";
import { createObjectSet } from "./objectSet/createObjectSet.js";
import type { ObjectSetFactory } from "./objectSet/ObjectSetFactory.js";
import {
  createStandardOntologyProviderFactory,
  type OntologyCachingOptions,
} from "./ontology/StandardOntologyProvider.js";
import { USER_AGENT } from "./util/UserAgent.js";

/** @internal */
export function createMinimalClient(
  metadata: MinimalClientParams["metadata"],
  baseUrl: string,
  tokenProvider: () => Promise<string>,
  options: OntologyCachingOptions & { logger?: Logger } = {},
  fetchFn: (
    input: Request | URL | string,
    init?: RequestInit | undefined,
  ) => Promise<Response> = global.fetch,
  objectSetFactory: ObjectSetFactory<any, any> = createObjectSet,
) {
  if (process?.env?.NODE_ENV !== "production") {
    try {
      new URL(baseUrl);
    } catch (e) {
      const hint =
        !baseUrl.startsWith("http://") || !baseUrl.startsWith("https://")
          ? ". Did you forget to add 'http://' or 'https://'?"
          : "";
      throw new Error(`Invalid stack URL: ${baseUrl}${hint}`);
    }
  }

  const minimalClient: MinimalClient = {
    ...createSharedClientContext(
      baseUrl,
      tokenProvider,
      USER_AGENT,
      fetchFn,
    ),
    objectSetFactory,
    ontologyRid: metadata.ontologyRid,
    logger: options.logger,
    clientCacheKey: {} as ClientCacheKey,
    requestContext: {},
  } satisfies Omit<
    MinimalClient,
    "ontologyProvider"
  > as any;

  return Object.assign(minimalClient, {
    ontologyProvider: createStandardOntologyProviderFactory(
      options,
    )(minimalClient),
  });
}
