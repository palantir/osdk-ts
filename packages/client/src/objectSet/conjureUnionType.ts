/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

type DeepWriteable<T> = {
  -readonly [P in keyof T]: DeepWriteable<T[P]>;
};

/**
 * Helper function that creates the shape of a conjure union in fewer bytes than manually declaring it (when compressed)
 *
 * e.g `{type:"base",base:{foo:5}}` becomes `a("base",{foo:5})`
 */
export function conjureUnionType<T extends string, const V>(
  type: T,
  value: V,
): { type: T } & Record<T, DeepWriteable<V>> {
  return {
    type,
    [type]: value,
  } as { type: T } & Record<T, V>;
}
