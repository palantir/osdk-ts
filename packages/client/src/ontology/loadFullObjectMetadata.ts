/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ObjectTypeDefinition } from "@osdk/api";
import {
  getObjectTypeV2,
  listOutgoingLinkTypesV2,
} from "@osdk/foundry/OntologiesV2_ObjectTypeV2";
import { listInterfaceTypes } from "@osdk/foundry/OntologiesV2_OntologyObjectV2";
import type {
  LinkTypeSideV2,
  ListOutgoingLinkTypesResponseV2,
  ObjectTypeFullMetadata,
} from "@osdk/foundry/types";
import { wireObjectTypeFullMetadataToSdkObjectTypeDefinition } from "@osdk/generator-converters";
import type { ConjureContext } from "conjure-lite";
import invariant from "tiny-invariant";
import type { LoadAllOntologiesResponse } from "../generated/ontology-metadata/api/LoadAllOntologiesResponse.js";
import type { OntologyLoadEntitiesRequest } from "../generated/ontology-metadata/api/OntologyLoadEntitiesRequest.js";
import { loadAllOntologies } from "../generated/ontology-metadata/api/OntologyMetadataService/loadAllOntologies.js";
import { loadOntologyEntities } from "../generated/ontology-metadata/api/OntologyMetadataService/loadOntologyEntities.js";
import type { MinimalClient } from "../MinimalClientContext.js";

async function loadAllOutgoingLinkTypes(
  client: MinimalClient,
  objtype: string,
) {
  const linkTypes: LinkTypeSideV2[] = [];
  let pageToken: string | undefined;
  do {
    const result: ListOutgoingLinkTypesResponseV2 =
      await listOutgoingLinkTypesV2(
        client,
        client.ontologyRid,
        objtype,
        { pageToken },
      );

    pageToken = result.nextPageToken;
    linkTypes.push(...result.data);
  } while (pageToken != null);

  return linkTypes;
}

function makeConjureContext(
  client: MinimalClient,
  servicePath: string,
): ConjureContext {
  return {
    baseUrl: client.stack,
    servicePath,
    fetchFn: client.fetch,
    tokenProvider: async () => await client.tokenProvider(),
  };
}

export async function loadFullObjectMetadata(
  client: MinimalClient,
  objtype: string,
): Promise<ObjectTypeDefinition<any, any>> {
  const conjureCtx = makeConjureContext(client, "/ontology-metadata/api");

  const [objectType, linkTypes, interfaceTypes, metadata] = await Promise.all([
    getObjectTypeV2(
      client,
      client.ontologyRid,
      objtype,
    ),

    loadAllOutgoingLinkTypes(client, objtype),
    listInterfaceTypes(
      client,
      client.ontologyRid,
      { pageSize: 200, preview: true },
    ),
    await loadAllOntologies(conjureCtx, {}),
  ]);

  const sharedPropertyTypeMapping = await loadSptMap(
    conjureCtx,
    metadata,
    client.ontologyRid,
    objectType.rid,
  );

  const full: ObjectTypeFullMetadata = {
    implementsInterfaces: interfaceTypes.data.map(i => i.apiName),
    linkTypes,
    objectType,
    sharedPropertyTypeMapping,
  };

  // TODO: reuse the loaded interface data!
  return wireObjectTypeFullMetadataToSdkObjectTypeDefinition(full, true);
}

async function loadSptMap(
  ctx: ConjureContext,
  ontologyMetadata: LoadAllOntologiesResponse,
  ontologyRid: string,
  objectRid: string,
) {
  const ontologyVersion =
    ontologyMetadata.ontologies[ontologyRid].currentOntologyVersion;

  const body: OntologyLoadEntitiesRequest = {
    objectTypeVersions: {
      [objectRid]: ontologyVersion,
    },
    linkTypeVersions: {},
    loadRedacted: false,
    includeObjectTypesWithoutSearchableDatasources: true,
  };
  const entities = await loadOntologyEntities(ctx, body);
  const objectType = entities.objectTypes[objectRid];
  invariant(objectType, "object type should be loaded");

  const sptMap: Record<string, string> = {};
  for (const property of Object.values(objectType.propertyTypes)) {
    if (property.sharedPropertyTypeApiName && property.apiName) {
      sptMap[property.sharedPropertyTypeApiName] = property.apiName;
    }
  }

  return sptMap;
}
