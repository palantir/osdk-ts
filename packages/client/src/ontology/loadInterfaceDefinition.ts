/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { InterfaceDefinition } from "@osdk/api";
import { __UNSTABLE_wireInterfaceTypeV2ToSdkObjectDefinition } from "@osdk/generator-converters";
import { OntologiesV2 } from "@osdk/internal.foundry";
import type { MinimalClient } from "../MinimalClientContext.js";

export async function loadInterfaceDefinition(
  client: MinimalClient,
  objectType: string,
): Promise<InterfaceDefinition<any, any>> {
  const r = await OntologiesV2.OntologyInterfaces.getInterfaceType(
    client,
    await client.ontologyRid,
    objectType,
    { preview: true },
  );

  return __UNSTABLE_wireInterfaceTypeV2ToSdkObjectDefinition(r, true);
}
