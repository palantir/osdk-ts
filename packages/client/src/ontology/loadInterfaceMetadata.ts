/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { InterfaceMetadata } from "@osdk/api";
import { __UNSTABLE_wireInterfaceTypeV2ToSdkObjectDefinition } from "@osdk/generator-converters";
import * as OntologiesV2 from "@osdk/internal.foundry.ontologiesv2";
import type { MinimalClient } from "../MinimalClientContext.js";

export async function loadInterfaceMetadata(
  client: MinimalClient,
  objectType: string,
): Promise<InterfaceMetadata<any, any>> {
  const r = await OntologiesV2.OntologyInterfaces.getInterfaceType(
    client,
    await client.ontologyRid,
    objectType,
    { preview: true },
  );

  return __UNSTABLE_wireInterfaceTypeV2ToSdkObjectDefinition(r, true);
}
