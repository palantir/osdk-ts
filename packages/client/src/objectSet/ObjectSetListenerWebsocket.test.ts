/*
 * Copyright 2024 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ObjectOrInterfaceDefinition } from "@osdk/api";
import { $ontologyRid, Employee, Office } from "@osdk/client.test.ontology";
import type {
  ObjectSetSubscribeRequests,
  StreamMessage_subscribeResponses,
} from "@osdk/client.unstable.osw";
import { apiServer } from "@osdk/shared.test";
import ImportedWebSocket from "isomorphic-ws";
import { http, HttpResponse } from "msw";
import type { Logger } from "pino";
import type { MockedClass, MockedFunction, MockedObject } from "vitest";
import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest";
import { z } from "zod";
import { createMinimalClient } from "../createMinimalClient.js";
import type { MinimalClient } from "../MinimalClientContext.js";
import { conjureUnionType } from "./conjureUnionType.js";
import type { ObjectSetListener } from "./ObjectSetListener.js";
import { ObjectSetListenerWebsocket } from "./ObjectSetListenerWebsocket.js";

// it needs to be hoisted because its referenced from our mocked WebSocket
// which must be hoisted to work
const rootLogger = await vi.hoisted(async (): Promise<Logger> => {
  const pino = (await import("pino")).pino;
  const pinoPretty = await import("pino-pretty");
  const { EventEmitter } = await import("node:events");
  class PinoConsoleLogDestination extends EventEmitter {
    write(a: string) {
      // remove trailing newline since console.log adds one
      if (a.at(-1) === "\n") a = a.slice(0, -1);

      // This lets the test framework aggregate the logs per test, whereas direct to stdout does not
      console.log(a);
    }
  }
  return pino(
    { level: "info" },
    (pinoPretty.build)({
      sync: true,
      timestampKey: undefined,
      errorLikeObjectKeys: ["error", "err", "exception"],
      errorProps: "stack,cause,properties",
      ignore: "time,hostname,pid",
      destination: new PinoConsoleLogDestination(),
    }),
  );
});

// make local uses of WebSocket typed right
const MockedWebSocket = ImportedWebSocket as unknown as MockedWebSocket;

const STACK = "https://stack.palantir.com";

vi.mock("isomorphic-ws", async (importOriginal) => {
  const original = await importOriginal<
    { default: WebSocket }
  >();

  const WebSocket = createMockWebSocketConstructor(
    original.default,
    rootLogger,
  );
  return { default: WebSocket, WebSocket };
});

describe("ObjectSetListenerWebsocket", async () => {
  beforeAll(async () => {
    apiServer.listen();
    addLoggerToApiServer(rootLogger);
  });

  afterAll(() => {
    apiServer.close();
  });

  describe("basic setup", () => {
    let minimalClient: MinimalClient;

    const MINIMUM_RECONNECT_DELAY = 2_000;
    const OBJECT_SET_EXPIRY_MS = 60_000;

    let client: ObjectSetListenerWebsocket;
    let listener: MockedObject<
      Required<ObjectSetListener<ObjectOrInterfaceDefinition>>
    >;
    let oslwInst = 0;

    beforeEach(() => {
      minimalClient = createMinimalClient(
        { ontologyRid: $ontologyRid },
        STACK,
        async () => "myAccessToken",
        { logger: rootLogger },
      );

      client = new ObjectSetListenerWebsocket({
        ...minimalClient,
        logger: rootLogger.child({ oslwInst: oslwInst++ }),
      }, {
        minimumReconnectDelayMs: MINIMUM_RECONNECT_DELAY,
        objectSetExpiryMs: OBJECT_SET_EXPIRY_MS,
      });

      listener = {
        onChange: vi.fn(),
        onError: vi.fn(),
        onOutOfDate: vi.fn(),
      };

      let objectSetRidCounter = 0;
      apiServer.use(
        http.post(
          `${STACK}/object-set-service/api/objectSets/temporary`,
          () =>
            HttpResponse.json({
              objectSetRid: `rid.hi.${objectSetRidCounter++}`,
            }),
        ),
        http.post(
          `${STACK}/object-set-watcher/api/object-set-watcher/batchEnableWatcher`,
          async () => {
            return HttpResponse.json({});
          },
        ),
      );

      vi.useFakeTimers();
    });

    afterEach(() => {
      vi.restoreAllMocks();
      apiServer.resetHandlers();
    });

    it("should not send onOutOfDate since it has not subscribed yet", () => {
      expect(MockedWebSocket).not.toHaveBeenCalled();
      expect(listener.onOutOfDate).not.toHaveBeenCalled();
      expect(listener.onChange).not.toHaveBeenCalled();
      expect(listener.onError).not.toHaveBeenCalled();
    });

    describe("requests subscription", () => {
      let ws: MockedWebSocket;
      let unsubscribe: () => void;
      let subReq1: ObjectSetSubscribeRequests;

      beforeEach(async () => {
        [ws, unsubscribe] = await subscribeAndExpectWebSocket(
          client,
          listener,
        );

        subReq1 = await expectSingleSubscribeMessage(ws);
      });

      afterEach(() => {
        // just ensure we ended in good states, as we clear these as we consume them
        expect(MockedWebSocket).not.toHaveBeenCalled();

        // make sure we are unsubscribed so closing the socket doesn't trigger reconnects
        unsubscribe();

        // disconnect the socket
        setWebSocketState(ws, "close");

        vi.runAllTicks();

        // Make sure we cleaned up
        expectEqualRemoveAndAddListeners(ws);
      });

      it("only sends a single request", async () => {
        respondSuccessToSubscribe(ws, subReq1);
        // actually this is broken FIXME
        unsubscribe();
        expect(ws.send).not.toHaveBeenCalled();
      });

      describe("socket closed before subscription confirmed", () => {
        beforeEach(() => {
          setWebSocketState(ws, "close");
          expectEqualRemoveAndAddListeners(ws);
        });

        describe("reconnect", async () => {
          beforeEach(async () => {
            [ws] = await Promise.all([
              expectWebSocketConstructed(),
              // delay for connection reconnect
              vi.advanceTimersByTimeAsync(MINIMUM_RECONNECT_DELAY),
            ]);
            setWebSocketState(ws, "open");
          });

          describe("subscribe and respond", () => {
            beforeEach(async () => {
              const subReq2 = await expectSingleSubscribeMessage(ws);
              respondSuccessToSubscribe(ws, subReq2);
            });

            it("should not send out of date", () => {
              expect(listener.onOutOfDate).not.toHaveBeenCalled();
            });
          });
        });
      });

      describe("successfully subscribed", () => {
        beforeEach(() => {
          respondSuccessToSubscribe(ws, subReq1);
        });

        describe("additional subscription", async () => {
          let unsubscribe2: () => void;
          let subReq2: ObjectSetSubscribeRequests;
          beforeEach(async () => {
            [unsubscribe2, subReq2] = await Promise.all([
              client.subscribe({
                type: "base",
                objectType: Office.apiName,
              }, listener),

              expectSingleSubscribeMessage(ws),
            ]);
            rootLogger.fatal({ subReq2 });

            respondSuccessToSubscribe(ws, subReq2);
          });

          afterEach(() => {
            unsubscribe2();
          });

          it("does not trigger an out of date ", () => {
            expect(listener.onOutOfDate).not.toHaveBeenCalled();
          });
        });

        describe("temporary object set expires", () => {
          beforeEach(async () => {
            await vi.advanceTimersByTimeAsync(OBJECT_SET_EXPIRY_MS + 1000);
          });

          it("should resubscribe and fire out of date", async () => {
            vi.runAllTicks();
            const subReq2 = await expectSingleSubscribeMessage(ws);
            expect(subReq2.id).not.toEqual(subReq1.id);
            respondSuccessToSubscribe(ws, subReq2);

            expect(listener.onOutOfDate).toHaveBeenCalledTimes(1);
          });
        });

        describe("socket closed abruptly", () => {
          beforeEach(() => {
            setWebSocketState(ws, "close");
            expectEqualRemoveAndAddListeners(ws);
          });

          describe("reconnect, resubscribe successfully", async () => {
            beforeEach(async () => {
              [ws] = await Promise.all([
                expectWebSocketConstructed(),
                // delay for connection reconnect
                vi.advanceTimersByTimeAsync(MINIMUM_RECONNECT_DELAY),
              ]);
              setWebSocketState(ws, "open");

              const subReq2 = await expectSingleSubscribeMessage(ws);
              respondSuccessToSubscribe(ws, subReq2);
            });

            it("should send out of date", () => {
              expect(listener.onOutOfDate).toHaveBeenCalledTimes(1);
            });
          });
        });

        it("should not send onOutOfDate for first call", () => {
          expect(MockedWebSocket).not.toHaveBeenCalled();
          expect(listener.onOutOfDate).not.toHaveBeenCalled();
          expect(listener.onChange).not.toHaveBeenCalled();
          expect(listener.onError).not.toHaveBeenCalled();
        });
      });
    });
  });
});

interface RawWebSocketPlus
  extends Pick<ImportedWebSocket, "addEventListener" | "removeEventListener">
{
  _eventEmitter: EventTarget;
  readyState: 0 | 1 | 2 | 3;
  send: MockedFunction<ImportedWebSocket["send"]>;
  close: MockedFunction<ImportedWebSocket["close"]>;
}

interface MockedWebSocket
  extends MockedClass<typeof ImportedWebSocket>, MockedObject<RawWebSocketPlus>
{
}

type MockedListener = MockedObject<
  Required<ObjectSetListener<ObjectOrInterfaceDefinition>>
>;

function respondSuccessToSubscribe(
  ws: MockedWebSocket,
  subReq2: ObjectSetSubscribeRequests,
) {
  sendToClient<StreamMessage_subscribeResponses>(
    ws,
    conjureUnionType("subscribeResponses", {
      id: subReq2.id,
      responses: subReq2.requests.map((_, i) =>
        conjureUnionType("success", { id: `id-${i}` })
      ),
    }),
  );
}

function expectEqualRemoveAndAddListeners(ws: MockedWebSocket) {
  expect(ws.removeEventListener).toHaveBeenCalledTimes(
    ws.addEventListener.mock.calls.length,
  );
}

async function expectSingleSubscribeMessage(
  ws: MockedWebSocket,
): Promise<ObjectSetSubscribeRequests> {
  return SubscribeMessage.parse(
    await consumeSingleSend(ws),
  ) as ObjectSetSubscribeRequests;
}

async function subscribeAndExpectWebSocket(
  client: ObjectSetListenerWebsocket,
  listener: MockedListener,
): Promise<readonly [MockedWebSocket, () => void]> {
  const [ws, unsubscribe] = await Promise.all([
    expectWebSocketConstructed(),
    client.subscribe({
      type: "base",
      objectType: Employee.apiName,
    }, listener),
  ]);

  // otherwise the ObjectSetListenerWebSocket is sitting waiting for it to "connect"
  setWebSocketState(ws, "open");

  return [ws, unsubscribe] as const;
}

async function expectWebSocketConstructed(): Promise<MockedWebSocket> {
  expect(MockedWebSocket).toHaveBeenCalledTimes(0);

  const ws = await vi.waitFor(() => {
    expect(MockedWebSocket).toHaveBeenCalled();
    expect(MockedWebSocket.mock.results[0].type === "return");
    return MockedWebSocket.mock.results[0].value as MockedWebSocket;
  });

  expect(MockedWebSocket).toHaveBeenCalledTimes(1);
  MockedWebSocket.mockClear();
  expect(MockedWebSocket).toHaveBeenCalledTimes(0);

  return ws;
}

async function consumeSingleSend(ws: any) {
  return await vi.waitFor(() => {
    expect(ws.send).toBeCalledTimes(1);
    const result = JSON.parse(ws.send.mock.lastCall![0].toString());
    ws.send.mockClear();
    return result;
  });
}

function createMockWebSocketConstructor(
  OriginalWebSocket: WebSocket,
  logger: Logger,
): MockedWebSocket {
  let i = 0;
  const ret = vi.fn(function(..._args: any[]): MockedWebSocket {
    const webSocketInst = i++;
    logger.debug("WebSocket constructor called");
    const eventEmitter = new EventTarget();

    return {
      addEventListener: vi.fn(
        eventEmitter.addEventListener.bind(eventEmitter),
      ) as any,
      removeEventListener: vi.fn(
        eventEmitter.removeEventListener.bind(eventEmitter),
      ) as any,

      send: vi.fn((a, _b: any) => {
        logger.debug(
          { message: JSON.parse(a.toString()), webSocketInst },
          "send() called",
        );
      }),
      close: vi.fn(),
      _eventEmitter: eventEmitter,
      readyState: OriginalWebSocket.CONNECTING,
    } satisfies RawWebSocketPlus as any as MockedWebSocket;
    // ^ we only implement some things but the type system wants to think its the full deal,
    // thus the satisfies plus the cast
  }) as any as MockedWebSocket;

  Object.assign(ret, {
    OPEN: OriginalWebSocket.OPEN,
    CLOSED: OriginalWebSocket.CLOSED,
    CLOSING: OriginalWebSocket.CLOSING,
    CONNECTING: OriginalWebSocket.CONNECTING,
  });

  return ret;
}

function setWebSocketState(ws: MockedWebSocket, readyState: "open" | "close") {
  const newState = readyState === "open"
    ? ImportedWebSocket.OPEN
    : ImportedWebSocket.CLOSED;

  if (newState === ws.readyState) return;

  ws.readyState = newState;
  ws._eventEmitter.dispatchEvent(new Event(readyState, {}));
}

function addLoggerToApiServer(logger: Logger) {
  const z = (
    name: string,
    { requestId, request }: { requestId: string; request: Request },
  ) => logger.trace({ requestId, url: request.url }, name);

  const eventNames = [
    "request:start",
    "request:match",
    "request:end",
    "request:unhandled",
    "response:bypass",
    "response:mocked",
  ] as const;

  for (const c of eventNames) {
    apiServer.events.on(c as typeof eventNames[number], z.bind(undefined, c));
  }
}

const SubscribeMessage = z.object({
  id: z.string(),
  requests: z.array(z.object({
    objectSet: z.string(),
    objectSetContext: z.any(),
    watchAllLinks: z.boolean(),
  })),
});

class MessageEvent extends Event {
  data: string;
  constructor(data: any) {
    super("message");
    this.data = JSON.stringify(data);
  }
}

function sendToClient<T>(ws: MockedWebSocket, t: T) {
  ws._eventEmitter.dispatchEvent(new MessageEvent(t));
}
