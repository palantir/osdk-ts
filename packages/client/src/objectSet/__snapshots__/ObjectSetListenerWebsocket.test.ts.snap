// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ObjectSetListenerWebsocket > basic setup > requests subscription > successfully subscribed > should correctly return reference update 1`] = `
{
  "object": {
    "$apiName": "Employee",
    "$metadata": {
      "apiName": "Employee",
      "description": "A full-time or part-time 

 employee of our firm",
      "displayName": "Employee",
      "icon": {
        "color": "blue",
        "name": "person",
        "type": "blueprint",
      },
      "implements": [
        "FooInterface",
      ],
      "interfaceMap": {
        "FooInterface": {
          "fooSpt": "fullName",
        },
      },
      "inverseInterfaceMap": {
        "FooInterface": {
          "fullName": "fooSpt",
        },
      },
      "links": {
        "lead": {
          "multiplicity": false,
          "targetType": "Employee",
        },
        "officeLink": {
          "multiplicity": false,
          "targetType": "Office",
        },
        "peeps": {
          "multiplicity": true,
          "targetType": "Employee",
        },
      },
      "pluralDisplayName": "Employees",
      "primaryKeyApiName": "employeeId",
      "primaryKeyType": "integer",
      "properties": {
        "class": {
          "description": "",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "string",
        },
        "employeeId": {
          "description": undefined,
          "displayName": undefined,
          "multiplicity": false,
          "nullable": false,
          "type": "integer",
        },
        "employeeLocation": {
          "description": "Geotime series reference of the location of the employee",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "geotimeSeriesReference",
        },
        "employeeSensor": {
          "description": "TimeSeries sensor of the status of the employee",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "sensorTimeseries",
        },
        "employeeStatus": {
          "description": "TimeSeries of the status of the employee",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "stringTimeseries",
        },
        "fullName": {
          "description": undefined,
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "string",
        },
        "office": {
          "description": "The unique "ID" of the employee's \\"primary\\" assigned office.
 This is some more text.",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "string",
        },
        "startDate": {
          "description": "The date the employee was hired (most recently, if they were re-hired)",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "datetime",
        },
      },
      "rid": "ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44",
      "status": "ACTIVE",
      "titleProperty": "fullName",
      "type": "object",
      "visibility": "NORMAL",
      Symbol(InterfaceDefinitions): {
        "FooInterface": {
          "def": {
            "apiName": "FooInterface",
            "description": "Interface for Foo",
            "displayName": "Foo Interface",
            "implementedBy": [
              "Employee",
              "Person",
            ],
            "implements": [],
            "links": {},
            "properties": {
              "fooSpt": {
                "description": "A foo",
                "displayName": "Foo",
                "multiplicity": false,
                "nullable": true,
                "type": "string",
              },
            },
            "rid": "ri.interface.main.interface.1",
            "type": "interface",
          },
          "handler": undefined,
        },
      },
    },
    "$objectSpecifier": "Employee:12345",
    "$objectType": "Employee",
    "$primaryKey": "12345",
    "$title": undefined,
    "employeeId": "12345",
    "employeeLocation": GeotimeSeriesPropertyImpl {
      "lastFetchedValue": {
        "time": "111",
        "value": {
          "coordinates": [
            100,
            200,
          ],
          "type": "Point",
        },
      },
    },
  },
  "state": "ADDED_OR_UPDATED",
}
`;

exports[`ObjectSetListenerWebsocket > basic setup > requests subscription > successfully subscribed > should correctly return regular updates 1`] = `
{
  "object": {
    "$apiName": "Employee",
    "$metadata": {
      "apiName": "Employee",
      "description": "A full-time or part-time 

 employee of our firm",
      "displayName": "Employee",
      "icon": {
        "color": "blue",
        "name": "person",
        "type": "blueprint",
      },
      "implements": [
        "FooInterface",
      ],
      "interfaceMap": {
        "FooInterface": {
          "fooSpt": "fullName",
        },
      },
      "inverseInterfaceMap": {
        "FooInterface": {
          "fullName": "fooSpt",
        },
      },
      "links": {
        "lead": {
          "multiplicity": false,
          "targetType": "Employee",
        },
        "officeLink": {
          "multiplicity": false,
          "targetType": "Office",
        },
        "peeps": {
          "multiplicity": true,
          "targetType": "Employee",
        },
      },
      "pluralDisplayName": "Employees",
      "primaryKeyApiName": "employeeId",
      "primaryKeyType": "integer",
      "properties": {
        "class": {
          "description": "",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "string",
        },
        "employeeId": {
          "description": undefined,
          "displayName": undefined,
          "multiplicity": false,
          "nullable": false,
          "type": "integer",
        },
        "employeeLocation": {
          "description": "Geotime series reference of the location of the employee",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "geotimeSeriesReference",
        },
        "employeeSensor": {
          "description": "TimeSeries sensor of the status of the employee",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "sensorTimeseries",
        },
        "employeeStatus": {
          "description": "TimeSeries of the status of the employee",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "stringTimeseries",
        },
        "fullName": {
          "description": undefined,
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "string",
        },
        "office": {
          "description": "The unique "ID" of the employee's \\"primary\\" assigned office.
 This is some more text.",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "string",
        },
        "startDate": {
          "description": "The date the employee was hired (most recently, if they were re-hired)",
          "displayName": undefined,
          "multiplicity": false,
          "nullable": true,
          "type": "datetime",
        },
      },
      "rid": "ri.ontology.main.object-type.401ac022-89eb-4591-8b7e-0a912b9efb44",
      "status": "ACTIVE",
      "titleProperty": "fullName",
      "type": "object",
      "visibility": "NORMAL",
      Symbol(InterfaceDefinitions): {
        "FooInterface": {
          "def": {
            "apiName": "FooInterface",
            "description": "Interface for Foo",
            "displayName": "Foo Interface",
            "implementedBy": [
              "Employee",
              "Person",
            ],
            "implements": [],
            "links": {},
            "properties": {
              "fooSpt": {
                "description": "A foo",
                "displayName": "Foo",
                "multiplicity": false,
                "nullable": true,
                "type": "string",
              },
            },
            "rid": "ri.interface.main.interface.1",
            "type": "interface",
          },
          "handler": undefined,
        },
      },
    },
    "$objectSpecifier": "Employee:undefined",
    "$objectType": "Employee",
    "$primaryKey": undefined,
    "$title": undefined,
    "employeeId": 1,
  },
  "state": "ADDED_OR_UPDATED",
}
`;
